### PR TITLE
feat: thorchain lp primitives in @vultisig/core-chain

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Summary

Asym RUNE-side THORChain LP primitives, lifted to a shared `@vultisig/core-chain` module so future consumers (mcp-ts, vultisig-windows, the agent app once it migrates to the SDK) all read from one source of truth instead of hand-rolling memo formats.

**Stacked PR**: vultisig/mcp-ts#12 wires these primitives into 4 MCP tools (`query_thorchain_pools`, `build_thorchain_lp_add`, `build_thorchain_lp_remove`, `query_thorchain_lp_position`) and validates them end-to-end with curl against live mainnet midgard + thornode.

This PR ships only the **building blocks**. v1 is intentionally narrow: asym RUNE-side adds + removes, no asset-side, no explicit symmetric pairing.

## What's in the box

New module at `packages/core/chain/chains/cosmos/thor/lp/`:

- **`affiliate.ts`** — `VULTISIG_AFFILIATE_NAME = 'vi'`, `VULTISIG_AFFILIATE_LP_BPS = 0`. Sourced from `vultisig-ios/.../THORChainSwaps.swift:27` and confirmed against the production memo fixtures (`TestData/thorchainswap.json` carries `:vi:0`).
- **`memo.ts`** — `addLpMemo({ pool, pairedAddress?, affiliate?, affiliateBps? })` and `removeLpMemo({ pool, basisPoints })`. For the asym RUNE-side default the memo looks like `+:BTC.BTC::vi:0` — empty paired-address slot is intentional, that's how THORChain registers the asymmetric position. Withdraws don't include affiliate per spec.
- **`payload.ts`** — `buildThorchainLpAddPayload` / `buildThorchainLpRemovePayload`. Returns flat `ThorchainLpAddPayload` / `ThorchainLpRemovePayload` shapes that flow through the agent backend's `tx_ready` SSE event. Single nesting level on purpose — the 2026-04-09 audit flagged tool result flattening as a known wire-level hazard for nested fields. Uses `0.02 RUNE` dust for removes (matches `vultisig-windows/core/ui/vault/deposit/keysignPayload/build.ts:325`).
- **`pools.ts`** — `getThorchainPools()` over Midgard `/v2/pools`, available-only by default. `THORCHAIN_MIDGARD_BASE_URL` constant exported.
- **`position.ts`** — `getThorchainLpPosition({ thorAddress, pool })` over Midgard `/v2/member/{addr}`. Returns `null` on 404 (expected for fresh addresses) or when the address has no position in the requested pool. Computes `isPending` from `runePending`/`assetPending`.
- **`validation.ts`** — `assertPoolDepositable(pool)` fail-fast gate over thornode `/thorchain/pool/{asset}`. Throws on anything other than `Available`.
- **`index.ts`** — barrel re-export for the whole module.

Plus 7 new entries in `packages/core/chain/package.json` exports (one per file + the barrel).

## Tests

22 unit tests, all passing locally:

```
✓ packages/core/chain/chains/cosmos/thor/lp/memo.test.ts (12 tests)
✓ packages/core/chain/chains/cosmos/thor/lp/payload.test.ts (10 tests)
```

Pure-function tests only — no fetch in the test path. Pools/position/validation are thin wrappers over `@vultisig/lib-utils/queryUrl`, matching the existing thor sibling style (`getThorchainInboundAddress`, `getThorNetworkInfo`).

## Risk

Low. New module, no consumers yet, no migration of existing code. Build is clean (`yarn build:all` ran end-to-end) so I haven't broken any downstream sdk/rujira consumers.

The one thing worth a second look: the `vi` affiliate THORName is registered on mainnet via the iOS app already, but if it's somehow not, removes still work and adds will succeed (memo just won't be tracked in Midgard's affiliate stats). Not a correctness bug.

## What's next (separate PRs, not this one)

- `vultisig-mcp-ts`: 4 MCP tools wrapping these primitives (`query_thorchain_pools`, `build_thorchain_lp_add`, `build_thorchain_lp_remove`, `query_thorchain_lp_position`). Already in flight.
- `vultisig-agent-backend`: register the tools, system prompt teaching, emit `tx_ready` (deferred until app can sign).
- `vultiagent-poc`: THORChain MsgDeposit signing path. Blocked on VA-133 (SDK migration into the app).

## Test plan

- [x] `yarn workspace @vultisig/core-chain build` — clean
- [x] `yarn build:all` — clean (no downstream regressions)
- [x] `yarn vitest run packages/core/chain/chains/cosmos/thor/lp` — 22/22 passing
- [ ] CodeRabbit review

## Summary by CodeRabbit

* **New Features**
  * Added THORChain liquidity-pool support: add/remove liquidity, pool listing, user LP positions, depositability checks, default affiliate settings, and standardized payloads/memos.
  * SDK now exposes THORChain LP primitives at the top level.

* **Chores**
  * Improved HTTP fetch error reporting with a typed error object and updated package exports.

* **Tests**
  * Added comprehensive tests for memo generation, payload builders, pools, and validation.

### CODE REVIEW

File: packages/core/chain/chains/cosmos/thor/lp/pools.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+export const getThorchainPools = async (
+  options: GetThorchainPoolsOptions = {}
+): Promise<ThorchainPoolSummary[]> => {
+  const status = options.status === undefined ? 'available' : options.status
+  const url =
+    status === null
+      ? `${THORCHAIN_MIDGARD_BASE_URL}/v2/pools`
+      : `${THORCHAIN_MIDGARD_BASE_URL}/v2/pools?status=${encodeURIComponent(status)}`
+  const raw = await queryUrl<RawPool[]>(url)
+  return raw.map(normalizePool)
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/core/chain/chains/cosmos/thor/lp/pools.ts]: done @coderabbitai — added `Array.isArray` guard before `.map` so a non-array Midgard payload throws a clear error instead of TypeError-ing on undefined. fix in commit 2e1c78e9.

**Comment by gomesalexandre** [comment]
@coderabbitai dis tackled in 2e1c78e9. all 4 points addressed:

- **pools.ts (Major)**: \`Array.isArray\` guard before \`.map\` (replied inline)
- **pools.ts (nit)**: renamed \`THORCHAIN_MIDGARD_BASE_URL\` → \`thorchainMidgardBaseUrl\` to match the camelCase convention used by the rest of the thor module
- **position.ts (nit)**: hardened the 404 detection regex (now matches both \`HTTP 404 ...\` and \`... 404 Not Found ...\` patterns) and added a TODO comment pointing at \`assertFetchResponse\` as the proper fix. Couldn't switch to status-based detection without modifying \`@vultisig/lib-utils/queryUrl\`'s public API which feels like scope creep for this PR
- **validation.ts (nit)**: defensive type narrowing on the \`queryUrl\` result before reading \`.status\`, with a clear error if the response shape is unexpected

also folded in a small bonus: \`export *\` of the lp module from \`@vultisig/sdk\` so consumers can \`import { addLpMemo } from '@vultisig/sdk'\` without pulling \`@vultisig/core-chain\` explicitly.

22 unit tests still passing.

**Comment by gomesalexandre** [comment]
@coderabbitai dis tackled in 16145535 — added the same Array.isArray guard on raw.pools in position.ts. mirrors the pools.ts pattern.

**Comment by NeOMakinG** [comment]
## QA Report — feat: thorchain lp primitives in @vultisig/core-chain

**Tested:** 2026-04-07
**Branch:** feat_thor_lp_primitives
**Scope:** packages/core/chain/chains/cosmos/thor/lp/, packages/sdk/src/index.ts
**Environment:** local build + vitest

### Results

| Test | Status | Notes |
|------|--------|-------|
| `yarn workspace @vultisig/core-chain build` | PASS | Clean, no errors or warnings from LP files |
| `yarn build:all` | PASS | No downstream regressions across all packages |
| `yarn vitest run .../lp` — memo.test.ts | PASS | 12/12 tests pass |
| `yarn vitest run .../lp` — payload.test.ts | PASS | 10/10 tests pass |
| TypeScript (core-chain) | PASS | No errors in LP files |
| TypeScript (packages/sdk) | PRE-EXISTING ERRORS | FiatValueService/SwapService test TS errors exist on main — not introduced by this PR |
| Pool math correctness | PASS | See analysis below |
| Memo format correctness | PASS | `+:BTC.BTC::vi:0` matches THORChain spec for asym RUNE-side |
| Withdraw memo format | PASS | `-:POOL:BPS` — no affiliate suffix, correct per spec |
| Cross-repo impact | PASS | No consumer yet; `export *` in packages/sdk/src/index.ts is the only change to existing files |

### Evidence

22/22 vitest tests passing:

```
 ✓ packages/core/chain/chains/cosmos/thor/lp/memo.test.ts (12 tests) 2ms
 ✓ packages/core/chain/chains/cosmos/thor/lp/payload.test.ts (10 tests) 2ms

 Test Files  2 passed (2)
      Tests  22 passed (22)
   Duration  668ms
```

### Protocol Correctness Assessment

**Memo format — verified correct:**
- Asym RUNE-side add: `+:BTC.BTC::vi:0` — empty paired-address slot is intentional and required by THORChain to register an asymmetric position. Matches [THORChain memo spec](https://dev.thorchain.org/concepts/memos.html).
- Full withdraw: `-:BTC.BTC:10000` — correct, no affiliate suffix per spec.
- Partial withdraw: `-:ETH.ETH:2500` — correct.

**Dust amount for removes — verified correct:**
- `LP_REMOVE_DUST_RUNE_BASE_UNITS = '2000000'` = 0.02 RUNE (8 decimals). Matches vultisig-windows reference (`build.ts:325`). Sufficient to make MsgDeposit valid on-chain.

**Input validation — solid:**
- `removeLpMemo` rejects `basisPoints` outside `[1, 10000]`, non-integer, NaN. Tests cover all edge cases.
- `buildThorchainLpAddPayload` rejects non-positive or non-integer-string amounts via `isPositiveBaseUnitString`.

**Security — no fund-loss risk identified:**
- All functions are pure builders; no signing, broadcast, or key handling.
- Pool validation (`assertPoolDepositable`) is a fail-fast gate that hits thornode before payload construction — prevents adds to Staged/Suspended pools.
- `isMidgardNotFoundError` uses regex on the error message to detect 404s. This is a known fragility (documented with a TODO in the code) — if `queryUrl`/`assertFetchResponse` changes its error format, position lookups on fresh addresses could start throwing instead of returning null. Low risk for now, tracked.

### Minor Observations (non-blocking)

1. **`isMidgardNotFoundError` string-match fragility** — the PR itself documents this as a TODO. Worth a follow-up to add a typed `HttpError` to `@vultisig/lib-utils/assertFetchResponse`. Not blocking.

2. **No `pool` format validation** — `addLpMemo` and `removeLpMemo` accept any string as the `pool` parameter. A caller passing `"btc.btc"` (lowercase) instead of `"BTC.BTC"` would produce a memo that THORChain rejects silently. Consider a lightweight regex check (`/^[A-Z0-9\-]+\.[A-Z0-9\-]+(-0X[A-F0-9]+)?$/i`) or at least a note in the type doc. Not blocking for v1.

3. **`export *` in packages/sdk/src/index.ts** — flat re-export from `@vultisig/core-chain/chains/cosmos/thor/lp` brings all exported names into the top-level SDK namespace. If downstream consumers do `import { ... } from '@vultisig/sdk'`, name collisions are possible as the LP module grows. Worth watching, not a problem today.

### Recommendation

**PASS.** Build clean, all 22 unit tests green, protocol math verified correct, no security issues. The two observations above are non-blocking and the author has already self-documented the main one. Ready to merge.

**Review by NeOMakinG — APPROVED** [review]
Reviewed the full diff + ran the test suite locally. Clean ship.

**Build:** `yarn build:all` passes end-to-end, no regressions.
**Tests:** 22/22 vitest unit tests green (`memo.test.ts` 12/12, `payload.test.ts` 10/10).
**Protocol correctness:** Memo formats match the THORChain spec — empty paired-address slot in `+:BTC.BTC::vi:0` correctly signals asymmetric RUNE-side position; `-:POOL:BPS` withdraw omits affiliate per spec; 0.02 RUNE dust constant matches the vultisig-windows reference.
**Security:** Pure builders only, no signing/broadcast. `assertPoolDepositable` provides a solid fail-fast gate against adds to paused pools. No fund-loss risk.

Two minor non-blocking notes (detailed in QA comment above):
1. `isMidgardNotFoundError` string-match on error message — self-documented TODO, low risk.
2. No pool-id format validation — uppercase `BTC.BTC` is a contract by convention, not enforced. Fine for v1 given no consumers yet.

The `export *` re-export in `packages/sdk/src/index.ts` is the right call for discoverability now; can be scoped down if the namespace gets crowded later.

Overall: well-structured module with good boundaries, thorough tests, clear inline docs. Approve.

**Comment by gomesalexandre** [comment]
tackled in 539acd95 ser @NeoMakinG

- midgard 404 check is now a typed `HttpResponseError.status === 404` instead of string-matching `err.message`. new `HttpResponseError` class carries status/statusText/url/body so callers can branch on structured fields
- pool id format validation via `assertValidPoolId` / `isValidPoolId` in `pools.ts`, wired into `addLpMemo` / `removeLpMemo` / `assertPoolDepositable` / `getThorchainLpPosition` so malformed ids fail fast at public entry points. covers `CHAIN.ASSET` + `CHAIN.ASSET-CONTRACT`, rejects lowercase, mixed case, wrong separators, whitespace etc. 28 parameterized tests in `pools.test.ts`

`yarn test:core` green (63/63), `yarn typecheck` clean, `yarn build:all` passes
